src/contract.rs

```diff
@@ -964,8 +1056,25 @@ fn try_burn<S: Storage, A: Api, Q: Querier>(
     env: Env,
     amount: Uint128,
 ) -> StdResult<HandleResponse> {
-    let sender_address = deps.api.canonical_address(&env.message.sender)?;
+    let mut config = Config::from_storage(&mut deps.storage);
+    let constants = config.constants()?;
+    if !constants.burn_is_enabled {
+        return Err(StdError::generic_err(
+            "Burn functionality is not enabled for this token.",
+        ));
+    }
     let amount = amount.u128();
+    let mut total_supply = config.total_supply();
+    if let Some(new_total_supply) = total_supply.checked_sub(amount) {
+        total_supply = new_total_supply;
+    } else {
+        return Err(StdError::generic_err(
+            "You're trying to burn more than is available in the total supply",
+        ));
+    }
+    config.set_total_supply(total_supply);
```

Comment:
like in redeem, we should check the user's balance before checking the total supply